### ISSUE TITLE
virgo 

### ISSUE THREAD

**Issue by iu4pej** [0xCoto/Virgo#6]
hello my name is gianni my city is ravenna on the adriatic coast, sorry for the inconvenience i downloaded virgo 3.6.3 i have this problem that i did not have with the previous version. i am using usrp b210, 3d corner reflector antenna. ubuntu 18.04. from the terminal I get this error in bandwidth 
All-Series:~/Virgo/virgo$ python  virgo.py -f 1420e6 -b 6e6 -c 1024 -t 1 -d 3600 -rf 55
gr-osmosdr 0.1.5 (0.1.5) gnuradio 3.7.14.0
built-in source types: file fcd rtl_tcp uhd rfspace redpitaya 
[INFO] [UHD] linux; GNU C++ version 7.5.0; Boost_106501; UHD_3.14.0.HEAD-0-g6875d061
[INFO] [B200] Detected Device: B210
[INFO] [B200] Operating over USB 3.
[INFO] [B200] Initialize CODEC control...
[INFO] [B200] Initialize Radio control...
[INFO] [B200] Performing register loopback test... 
[INFO] [B200] Register loopback test passed
[INFO] [B200] Performing register loopback test... 
[INFO] [B200] Register loopback test passed
[INFO] [B200] Setting master clock rate selection to 'automatic'.
[INFO] [B200] Asking for clock rate 16.000000 MHz... 
[INFO] [B200] Actually got clock rate 16.000000 MHz.
-- Using subdev spec 'A:A A:B'.
[INFO] [B200] Asking for clock rate 48.000000 MHz... 
[INFO] [B200] Actually got clock rate 48.000000 MHz.
[WARNING] [MULTI_USRP] Setting DC offset is not possible on this device.
[WARNING] [MULTI_USRP] Setting IQ balance is not possible on this device.
[WARNING] [AD936X] The requested bandwidth 0.000000 MHz is out of range (0.200000 - 56.000000 MHz).
The bandwidth has been forced to 0.200000 MHz.
in this situation I connect the sdr b210 to the rf generator, 0dbm and a frequency of 1420.00500mhz, I see the carrier in plot_png
and that's fine.
any dev_args I enter have the same error.
i ran a test with usrp n210 and it works fine.

**Comment by iu4pej** [comment]
i 0xCoto
do you have a suggestion for my problem?
thank you.

(wola-grc and ftf-grc in gnuradio works fine, I can set the bandwidth I want and I don't get any errors)

**Comment by 0xCoto** [comment]
Hi Gianni, can you show me the "wola-grc/ftf-grc in gnuradio" that looks fine (I assume you mean you've hardcoded some value in the flowchart)? I want to see what kind of parameters you've set at your SDR Source block, because it seems that the 6 MHz of bandwidth is not passed to the flowchart correctly (unlike the rest of the arguments).

**Comment by iu4pej** [comment]
![Schermata del 2021-03-07 17-14-19](https://user-images.githubusercontent.com/80112910/110246973-23ded880-7f6a-11eb-86c9-b4dc173f21cc.png)
![Schermata del 2021-03-07 17-16-38](https://user-images.githubusercontent.com/80112910/110246975-24776f00-7f6a-11eb-91e4-aaadf490fef1.png)
![Schermata del 2021-03-07 17-17-35](https://user-images.githubusercontent.com/80112910/110246976-24776f00-7f6a-11eb-91ec-5396fcbb13d2.png)
![Schermata del 2021-03-07 17-19-08](https://user-images.githubusercontent.com/80112910/110246978-25100580-7f6a-11eb-8529-d932dbd2d778.png)
![Schermata del 2021-03-07 17-19-24](https://user-images.githubusercontent.com/80112910/110246979-25a89c00-7f6a-11eb-8edf-ec98a815f82e.png)

**Comment by 0xCoto** [comment]
OK, I see. I typically replace the `Osmosdr Source` block with a `UHD: USRP Block` myself too to be honest, because it avoids problems with USRP devices (I've tried with both the USRP N210 and the B210 model). I'm pretty certain that it should be possible to interact with osmosdr in a reliable fashion with the USRPs, but I'm not quite sure why it has issues with the bandwidth. I've also been able to successfully adjust things like the master clock settings and the RX channels via the `dev_args` argument (Virgo/gr-osmosdr), so perhaps you could play around with that if you wihs to experiment.

But in general, I'd simply recommend replacing the osmo block with the UHD one if you're using a USRP device. I've found it to be the simplest and most reliable way to acquire data for me, so that's what I'd suggest to avoid running into issues.

**Comment by iu4pej** [comment]
i 0xCoto
thank you for help .
many spikes in my lab today !!!!


![iu4pej](https://user-images.githubusercontent.com/80112910/110680623-22ffae00-81d9-11eb-82e9-730ff9036c88.png)

**Comment by 0xCoto** [comment]
Very nice, looking good! A clear HI detection indeed.

Some suggestions to improve your results:
- Calibrate the spectrum, so the frequency response of the spectrum is flat and the hydrogen line is even more distinct;
- Use `f_rest=1420.4057517667e6` to better view the Doppler shift of the observed emissions (this will also add a secondary x-axis that displays the radial velocity of the source);
- Enable time-series median filtering (maybe on the frequency domain as well) to mitigate interference, and
- Mask the RFI channels on the right (around 1423 MHz). You can do that with `virgo.plot`'s `rfi` parameter.

You can apply the above tips like so (see [README](https://github.com/0xCoto/Virgo/blob/master/README.md#example-usage) for further examples & documentation):
```python
virgo.plot(obs_parameters=obs, n=20, m=35, f_rest=1420.4057517667e6,
           obs_file='observation.dat', cal_file='calibration.dat',
           rfi=[1422.5e6, 1423e6], plot_file='plot.png')
```

This should help filter out interference, and also calibrate your spectrum (make sure you've acquired a `calibration.dat` file by observing when the galactic plane is away from your telescope's beam).

**Comment by iu4pej** [comment]
Hello I did so, I directed the antenna on a cold sky area, and I acquired with wola.grc, in parameter, obs_file, value = calibration.dat. its ok ?
in the rf amplification chain i have a filter with a bandwidth of 5.5mhz with very steep sides, i am using a 3d corner reflector antenna.
![plot](https://user-images.githubusercontent.com/80112910/111902490-f6039480-8a3d-11eb-87f7-1275c17b9659.png)

**Comment by 0xCoto** [comment]
Try running an observation when the antenna is pointing to the galactic poles (so it is exposed to the galactic plane as little as possible, to avoid HI contamination), and save this file as `calibration.dat`. Don't set the duration super high, it must be:
- Short enough so the rotation of the Earth doesn't get the galactic plane in your beam (I can see your calibrated spectrum is slightly contaminated with HI), and
- Longer than your HI observation that will come next.

Then observe again **with the exact same parameters** (duration does not have to be the same), and the calibrated spectrum should be relatively flat (ideally).

Have you got the 3D corner reflector on a mount, or is it stationary? Aperture size?

**Comment by iu4pej** [comment]
hello thanks for your help, this antenna is very simple to build the project you can find in the yu1aw website (2.4 ghz 3d corner reflector), I have increased the side of the square to 3.8 lambda. at the moment the antenna is as shown in the photo       .
return loss is very good if i find my vna screenshot i can send it to you.
I will follow your advice regarding calibration.

![IMG_0964](https://user-images.githubusercontent.com/80112910/111921538-68ec2a00-8a95-11eb-985c-9daf0246a33a.JPG)

**Comment by iu4pej** [comment]
hi 0xCoto, I have refurbished some components of my 1.4ghz receiver, I have not used Virgo for a long time and most likely I have forgotten a lot of things. I am getting the following error. can you help me thanks. Gianni iu4pej

***@***.***:~$ cd Virgo/virgo
***@***.***:~/Virgo/virgo$ python virgo.py -f 1420e6 -b 6.25e6 -rf 65 -t 1 -c 1024 -d 120 -da uhd=a,addr=192.168.10.2,subdev=A:0 -if 0 -bb 0 -p iu4pej -r 1420.4057517667e6 -db --db -n 20 -m 35
Traceback (most recent call last):
File "virgo.py", line 1302, in <module>
main()
File "virgo.py", line 1294, in main
observe(obs_parameters=observation, obs_file=args.obs_file, start_in=args.start_in)
File "virgo.py", line 530, in observe
loc = obs_parameters['loc']
KeyError: 'loc'
***@***.***:~/Virgo/virgo$

**Comment by 0xCoto** [comment]
@kevinawilson Any ideas if the VLSR implementation makes command-line executions incompatible?

**Comment by iu4pej** [comment]
hi 0xCoto, I have refurbished some components of my 1.4ghz receiver, I have not used Virgo for a long time and most likely I have forgotten a lot of things. I am getting the following error. can you help me thanks. Gianni iu4pej

k@k-All-Series:~$ cd Virgo/virgo
k@k-All-Series:~/Virgo/virgo$ python virgo.py -f 1420e6 -b 6.25e6 -rf 65 -t 1 -c 1024 -d 120 -da uhd=a,addr=192.168.10.2,subdev=A:0 -if 0 -bb 0 -p iu4pej  -r 1420.4057517667e6   -db --db -n 20 -m 35 
Traceback (most recent call last):
  File "virgo.py", line 1302, in <module>
    main()
  File "virgo.py", line 1294, in main
    observe(obs_parameters=observation, obs_file=args.obs_file, start_in=args.start_in)
  File "virgo.py", line 530, in observe
    loc = obs_parameters['loc']
KeyError: 'loc'
k@k-All-Series:~/Virgo/virgo$

**Comment by iu4pej** [comment]
hi Apostolos ,

I apologize for the inconvenience I feel really stupid but I don't understand your answer.
should i create a cal.dat file with wola?
i reinstalled virgo in the new pc, gr-wola and gr-ftf work.
many thanks

**Comment by 0xCoto** [comment]
@iu4pej Regarding the KeyError: 'loc' issue, I pinged Kevin (a contributor to the code) because I suspect your error may be related to a recent addition of his that we haven't extensively tested.

Regarding calibration, sure - it's best to create a `cal.dat` file with WOLA (or FTF, if that's the spectrometer you've selected).

**Comment by 0xCoto** [comment]
@iu4pej I've just pushed a quick fix for command-line executions (hopefully it works without issues). Please try updating to **v3.8.1** using `pip install -U astro-virgo` and let me know!

**Comment by kevinawilson** [comment]
My apologies for the slow response. I'm away from home for Christmas.

I had not worked to make it compatible with the command-line execution, as I am not familiar with that aspect of the code. It looks like you have a fix in place. Do you need me to look into the issue?

**Comment by 0xCoto** [comment]
No worries Kevin, let's wait for iu4pej to check if the new release I just pushed did the trick for him and see if we need to make any changes.

**Comment by iu4pej** [comment]
i Apostolos

ok this work , all ok for usrp n210 , for usrp b210 i have this ,[WARNING] [AD936X] The requested bandwidth 0.000000 MHz is out of range (0.200000 - 56.000000 MHz).
The bandwidth has been forced to 0.200000 MHz., but for me not problem i use n210.

many thanks

**Comment by 0xCoto** [comment]
@iu4pej Could you share the line where you define your bandwidth? I've also tested with the USRP N210 with no issues, but the B210 should also work fine. You may have accidentally set a value in MHz (e.g. 1420) instead of Hz (1420e6).

**Comment by iu4pej** [comment]
![iu4pej](https://user-images.githubusercontent.com/80112910/151981164-b9fa4f43-4abd-4760-892a-0443357047a7.png)
